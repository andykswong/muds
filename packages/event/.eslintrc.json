{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "tsconfig.json",
    "sourceType": "module"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended"
  ],
  "env": {
    "es2022": true,
    "browser": true,
    "node": true,
    "jest": true
  },
  "ignorePatterns": [
  ],
  "rules": {
  }
}
