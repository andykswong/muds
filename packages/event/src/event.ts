/** Listener of events. */
export interface EventListener<T extends unknown[]> {
  (...data: T): void;
}

/** A typesafe event signal. */
export interface Event<T extends unknown[]> {
  /** Adds an event listener. */
  addListener(listener: EventListener<T>): void;

  /** Removes an event listener. */
  removeListener(listener: EventListener<T>): void;

  /** Emits an event. */
  emit(...data: T): void;
}

/** A typesafe event signal. */
export class Event<T> implements Event<T> {
  /** Creates a simple event instance. */
  public static create<T extends unknown[]>(): Event<T> {
    return new SimpleEvent();
  }
}

/** A simple Event implementation that triggers listeners immediately upon emitting. */
export class SimpleEvent<T extends unknown[]> implements Event<T> {
  private readonly listeners: EventListener<T>[] = [];

  public addListener(listener: EventListener<T>) {
    this.listeners.push(listener);
  }

  public removeListener(listener: EventListener<T>) {
    const index = this.listeners.indexOf(listener);
    if (index > -1) {
      this.listeners[index] = this.listeners[this.listeners.length - 1];
      this.listeners.pop();
    }
  }

  public emit(...data: T) {
    for (const listener of this.listeners) {
      listener(...data);
    }
  }
}
