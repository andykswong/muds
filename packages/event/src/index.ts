export * from './event';
export * from './multiplexer';
export * from './time';
