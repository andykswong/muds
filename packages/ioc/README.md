# @muds/ioc
<p align="center">
  <a href="https://www.npmjs.com/package/@muds/ioc"><img src="https://img.shields.io/npm/v/@muds/ioc?label=@muds/ioc" alt="NPM @muds/ioc" /></a>
  <a href="./LICENSE"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT" /></a>
</p>

## Overview
`@muds/ioc` is a minimal inversion of control (IoC) container library for [muds](https://github.com/andykswong/muds), a modular microframework for interactive data-oriented systems.

In `@muds/ioc`, the IoC container contents are defined by module classes, which declares provider functions for injectables. Modules are easily reusable, interchangeable and testable. Unlike many other JavaScript/TypeScript IoC libraries, `@muds/ioc` does not require you to add decorators nor make any change to the injectable classes, makes it easy to integrate with third-party libraries.

## API Documentation
http://andykswong.github.io/muds

## Usage
```shell
npm install --save @muds/ioc
```

The library requires the use of TypeScript [decorators](https://www.typescriptlang.org/docs/handbook/decorators.html), or [Babel legacy decorator](https://babeljs.io/docs/en/babel-plugin-proposal-decorators#legacy) if using pure JavaScript:
```javascript
import { Container, inject, module, multi, provide, singleton, tagged } from '@muds/ioc';

const id0 = Symbol('id0'), id1 = Symbol('id1'), id2 = Symbol('id2');

// 1. declare modules
@module()
class Module {
  // each provider method must be decorated with @provide(id). they can return any value
  @provide(id0)
  getFoo() { return 'foo'; }

  // singletons are cached once created
  // use @inject to inject a dependency by id
  @provide(id1) @singleton 
  getBar(@inject(id0) foo) {
    return { bar: foo };
  }

  // multiple providers with same id is supported
  // add tags using @tagged to differentiate them
  @provide(id1) @tagged({ type: 2 }) 
  getBar2() { return { bar: 2 }; }

  // use @multi to inject an array of injectables with same id
  @provide(id2)
  getBars(@inject(id1) @multi bars) { return bars; }
}

// 2. initialize IoC container and add modules
const container = Container.create();
container.add(new Module());

// 3. use get(id, tags?) or multiGet(id, tags?) to get injectables
container.get(id0); // 'foo'
container.multiGet(id1); // [{ bar: 'foo' }, { bar: 2 }]
container.get(id2); // [{ bar: 'foo' }, { bar: 2 }]
container.get(id1, { type: 2 }); // { bar: 2 }
```

## License
This repository and the code inside it is licensed under the MIT License. Read [LICENSE](./LICENSE) for more information.
