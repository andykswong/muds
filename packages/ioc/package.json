{
  "name": "@muds/ioc",
  "version": "0.1.2",
  "description": "Minimal inversion of control container for muds",
  "type": "module",
  "sideEffects": false,
  "main": "index.js",
  "types": "index.d.ts",
  "exports": {
    "import": "./index.js"
  },
  "files": [
    "/dist",
    "index.js",
    "index.d.ts"
  ],
  "scripts": {
    "prepublishOnly": "npm run clean && npm run build && npm test",
    "clean": "rimraf coverage docs dist",
    "prebuild": "npm run lint",
    "build": "npm run tsc && npm run babel",
    "lint": "eslint src --ext .ts,.tsx",
    "babel": "babel dist -d dist",
    "tsc": "tsc --project tsconfig.build.json",
    "test": "npm run jest",
    "jest": "cross-env NODE_OPTIONS=--experimental-vm-modules npx jest",
    "jest:watch": "npm run jest -- --watch",
    "doc": "typedoc --out docs src/index.ts"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/andykswong/muds.git"
  },
  "keywords": [
    "data-oriented",
    "microframework",
    "muds",
    "typescript",
    "di",
    "ioc",
    "dependency injection",
    "inversion of control"
  ],
  "author": "Andy K.S. Wong <andykswong@outlook.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/andykswong/muds/issues"
  },
  "homepage": "https://github.com/andykswong/muds",
  "devDependencies": {
    "@babel/cli": "^7.18",
    "@babel/core": "^7.18",
    "@babel/preset-env": "^7.18",
    "@types/jest": "^29.2",
    "@types/node": "^18.0",
    "@typescript-eslint/eslint-plugin": "^5.30",
    "@typescript-eslint/parser": "^5.30",
    "babel-plugin-add-import-extension": "^1.6",
    "copyfiles": "^2.4",
    "cross-env": "^7.0",
    "eslint": "^8.19",
    "jest": "^29.2",
    "reflect-metadata": "^0.1",
    "rimraf": "^3.0",
    "ts-jest": "^29.0",
    "typedoc": "^0.23",
    "typescript": "^4.7"
  },
  "jest": {
    "preset": "ts-jest/presets/default-esm",
    "extensionsToTreatAsEsm": [
      ".ts"
    ],
    "setupFilesAfterEnv": [
      "<rootDir>/jest.setup.ts"
    ],
    "testEnvironment": "node",
    "testMatch": [
      "**/__tests__/**/*.spec.ts"
    ],
    "collectCoverage": true,
    "collectCoverageFrom": [
      "src/**/*.ts",
      "!**/__tests__/**"
    ],
    "coverageDirectory": "coverage"
  },
  "typedoc": {
    "entryPoint": "src/index.ts"
  }
}
