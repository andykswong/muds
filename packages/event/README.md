# @muds/event
<p align="center">
  <a href="https://www.npmjs.com/package/@muds/event"><img src="https://img.shields.io/npm/v/@muds/event?label=@muds/event" alt="NPM @muds/event" /></a>
  <a href="./LICENSE"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT" /></a>
</p>

## Overview
`@muds/event` is the core event signal library for [muds](https://github.com/andykswong/muds), the modular microframework for interactive data-oriented systems. It provides the event emitter and listener interface, as well as several useful event implementations.

## API Documentation
http://andykswong.github.io/muds

## Usage
```shell
npm install --save @muds/event
```

```javascript
import { Event, EventMultiplexer } from '@muds/event';

const listener = (...msg) => console.log(...msg);
const event1 = Event.create(); // simple event
const event2 = Event.create();

event1.addListener(listener);
event1.emit('hello', 'world'); // log "hello world"
event1.removeListener(listener);

const multiplexer = new EventMultiplexer(); // multiplex events into a single event
multiplexer.addListener(listener);
multiplexer.add(event1, (event) => [1, ...event]); // add an event with event mapping
multiplexer.add(event2, (event) => [2, ...event]);
event1.emit('hello', 'again'); // log "1 hello again"
event2.emit('bye'); // log "2 bye"
```

```javascript
import { OnAnimationFrame } from '@muds/event';

const listener = (delta) => console.log(`delta: ${delta} ms`);
const event = new OnAnimationFrame(); // emit event with delta time on each requestAnimationFrame callback
event.addListener(listener);
event.start(); // log "delta: 16ms" periodically
// event.pause(); to stop
```

## License
This repository and the code inside it is licensed under the MIT License. Read [LICENSE](./LICENSE) for more information.
