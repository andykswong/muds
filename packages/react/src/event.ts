import { useState, useEffect, useRef } from 'react';
import { Event, EventListener } from '@muds/event';

/** Hook to listen to an event. */
export function useEventListener<E extends unknown[]>(
  event: Event<E>,
  listener: EventListener<E>,
): void {
  const savedListener = useRef<EventListener<E>>();

  useEffect(() => {
    savedListener.current = listener;
  }, [listener]);

  useEffect(() => {
    function onEvent(...event: E) {
      savedListener.current?.(...event);
    }
    event.addListener(onEvent);
    return () => event.removeListener(onEvent);
  }, [event]);
}

/** Hook to derive state from event. */
export function useEventReducer<S, E extends unknown[]>(
  event: Event<E>,
  reducer: (state: S, ...event: E) => S,
  initialState: S | (() => S),
): S {
  const [state, setState] = useState(initialState);
  const onEvent = useRef<EventListener<E>>();

  useEffect(() => {
    onEvent.current = (...event: E) => {
      setState(reducer(state, ...event));
    };
  }, [state, setState, reducer]);

  useEffect(() => {
    const listener = (...event: E) => onEvent.current?.(...event);
    event.addListener(listener);
    return () => event.removeListener(listener);
  }, [event]);

  return state;
}
