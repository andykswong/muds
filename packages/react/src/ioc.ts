import { createContext, useContext } from 'react';
import { Container, CustomTags, Identifier } from '@muds/ioc';

const ContainerContext = createContext(Container.create());

/** Provider of IoC container. */
export const ContainerProvider = ContainerContext.Provider;

/** Returns the current IoC container. */
export function useContainer(): Container {
  return useContext(ContainerContext);
}

/** Injects a value from IoC container. */
export function useInject<T>(id: Identifier, tags?: CustomTags): T | undefined {
  return useContainer().get(id, tags);
}

/** Injects values from IoC container. */
export function useMultiInject<T>(id: Identifier, tags?: CustomTags): T[] {
  return useContainer().getAll(id, tags);
}
