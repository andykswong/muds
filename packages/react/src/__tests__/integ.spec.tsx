import React from 'react';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { Event } from '@muds/event';
import { useEventReducer } from '../event';
import { ContainerProvider, useInject } from '../ioc';
import { Container, module, provide, singleton } from '@muds/ioc';


describe('integ', () => {
  const COUNTS = Symbol('counts'), INC_COUNT = Symbol('incCount');

  @module() class Module {
    @provide(COUNTS) @singleton counts() {
      return [10, 20];
    }

    @provide(INC_COUNT) @singleton onIncCount() {
      return Event.create<[number]>();
    }
  }

  function reducer(counts: number[], index: number) {
    counts[index]++;
    return [...counts];
  }

  function Counters() {
    const event = useInject(INC_COUNT) as Event<[number]>;
    const counts = useEventReducer(event, reducer, useInject(COUNTS) as number[]);
    return <span>{counts.map((count, i) => <button key={i} onClick={() => event.emit(i)}>{count}</button>)}</span>;
  }

  let container: Container;

  beforeEach(() => {
    container = Container.create();
    container.add(new Module());
  });

  test('integrate IoC container and events with React component', () => {
    let component!: ReactTestRenderer;

    act(() => {
      component = create(
        <ContainerProvider value={container}>
          <Counters />
        </ContainerProvider>
      );
    });

    expect(component.toJSON()).toMatchSnapshot();

    for (let i = 0; i < 2; ++i) {
      act(() => {
        const buttons = component.root.findAllByType('button');
        buttons[i].props.onClick();
      });
      expect(component.toJSON()).toMatchSnapshot();
    }

    act(() => component.unmount());
  });
});
