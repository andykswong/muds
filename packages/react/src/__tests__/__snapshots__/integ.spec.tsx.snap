// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integ integrate IoC container and events with React component 1`] = `
<span>
  <button
    onClick={[Function]}
  >
    10
  </button>
  <button
    onClick={[Function]}
  >
    20
  </button>
</span>
`;

exports[`integ integrate IoC container and events with React component 2`] = `
<span>
  <button
    onClick={[Function]}
  >
    11
  </button>
  <button
    onClick={[Function]}
  >
    20
  </button>
</span>
`;

exports[`integ integrate IoC container and events with React component 3`] = `
<span>
  <button
    onClick={[Function]}
  >
    11
  </button>
  <button
    onClick={[Function]}
  >
    21
  </button>
</span>
`;
